node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/HostCapacityMaintainer.java

```diff
@@ -77,99 +73,44 @@ public class DynamicProvisioningMaintainer extends NodeRepositoryMaintainer {
-            return;  // avoid removing excess hosts
+            return 0;  // avoid removing excess hosts
         }
 
-        excessHosts.forEach(host -> {
-            Optional<NodeMutex> optionalMutex = nodeRepository().nodes().lockAndGet(host, Duration.ofSeconds(10));
-            if (optionalMutex.isEmpty()) return;
-            try (NodeMutex mutex = optionalMutex.get()) {
-                if (host.state() != mutex.node().state()) return;
-                host = mutex.node();
-                // First mark the host as wantToDeprovision so that if hostProvisioner fails, this host
-                // * won't get new nodes allocated to it
-                // * will be selected as excess on next iteration of this maintainer
-                nodeRepository().nodes().deprovision(host.hostname(), Agent.DynamicProvisioningMaintainer, nodeRepository().clock().instant());
-                hostProvisioner.deprovision(host);
-                nodeRepository().nodes().removeRecursively(host, true);
-            } catch (UncheckedTimeoutException e) {
-                log.log(Level.WARNING, "Failed to deprovision " + host.hostname() +
-                                       ": Failed to get lock on node, will retry later");
-            } catch (RuntimeException e) {
-                log.log(Level.WARNING, "Failed to deprovision " + host.hostname() + ", will retry in " + interval(), e);
-            }
-        });
+        markForRemoval(excessHosts);
+        return 1;
     }
 
-    /** Replace the root disk of hosts that have requested soft-rebuild */
-    private void replaceRootDisk(NodeList nodes) {
-        NodeList softRebuildingHosts = nodes.rebuilding(true);
-        for (var host : softRebuildingHosts) {
-            Optional<NodeMutex> optionalMutex = nodeRepository().nodes().lockAndGet(host, Duration.ofSeconds(10));
-            if (optionalMutex.isEmpty()) return;
-            try (NodeMutex mutex = optionalMutex.get()) {
-                // Re-check flag while holding lock
-                host = mutex.node();
-                if (!host.status().wantToRebuild()) {
-                    continue;
+    private void markForRemoval(List<Node> excessHosts) {
+        if (excessHosts.isEmpty()) return;
+
+        try (var lock = nodeRepository().nodes().lockUnallocated()) {
+            NodeList nodes = nodeRepository().nodes().list(); // Reread nodes under lock
+            for (Node host : excessHosts) {
+                Optional<NodeMutex> optionalMutex = nodeRepository().nodes().lockAndGet(host, Duration.ofSeconds(10));
+                if (optionalMutex.isEmpty()) continue;
+                try (NodeMutex mutex = optionalMutex.get()) {
+                    host = mutex.node();
+                    if (!canRemoveHost(host)) continue;
+                    if (!nodes.childrenOf(host).stream().allMatch(HostCapacityMaintainer::canDeprovision))
+                        continue;
```

Comment:
These are linked to candidatesForRemoval --- perhaps there is a way for both to use the same function somehow?

Reply:
Done